packages/hydrogen/src/storefront.ts

```diff
@@ -432,20 +445,29 @@ function throwError<T>(
   response: Response,
   errors: StorefrontApiResponse<T>['errors'],
   ErrorConstructor = Error,
+  query?: string,
+  variables?: string,
 ) {
   const reqId = response.headers.get('x-request-id');
   const reqIdMessage = reqId ? ` - Request ID: ${reqId}` : '';
+  let graphiqlUrl = '';
+
+  if (query) {
+    graphiqlUrl = `\n\nTry running this query in isolotion to debug: http://localhost:3000/graphiql?query=${encodeURIComponent(
+      query,
+    )}${variables ? `&variables=${encodeURIComponent(variables)}` : ''}\n`;
+  }
 
   if (errors) {
     const errorMessages =
       typeof errors === 'string'
         ? errors
         : errors.map((error) => error.message).join('\n');
 
-    throw new ErrorConstructor(errorMessages + reqIdMessage);
+    throw new ErrorConstructor(errorMessages + reqIdMessage + graphiqlUrl);
   }
 
   throw new ErrorConstructor(
-    `API response error: ${response.status}` + reqIdMessage,
+    `API response error: ${response.status}` + reqIdMessage + graphiqlUrl,
   );
```

Comment:
Since `storefront.ts` doesn't have access to the local domain or port, what about simply adding the query/variables information to `error.cause` here, and then picking it up at the CLI level? ([around `lib/mini-oxygen.ts`](https://github.com/Shopify/hydrogen/blob/2023-04/packages/cli/src/lib/mini-oxygen.ts#L64))

```suggestion
  const reqId = response.headers.get('x-request-id');
  const errorMessage =
    (typeof errors === 'string'
      ? errors
      : errors?.map?.((error) => error.message).join('\n')) ||
    `API response error: ${response.status}`;

  throw new ErrorConstructor(
    errorMessage + (reqId ? ` - Request ID: ${reqId}` : ''),
    {
      cause: {
        errors,
        reqId,
        ...(process.env.NODE_ENV === 'development' && {query, variables}),
      },
    },
  );
```

Reply:
TIL `{ ...(false) }` evaluates to `{}`

Reply:
I mean, to get something useful in the production logs, everyone has to manually `try catch` their queries and mutations, which is really painful:

```ts
  let shop, product;
  try {
    const resp = await context.storefront.query<{
      product: ProductType & {selectedVariant?: ProductVariant};
      shop: Shop;
    }>(PRODUCT_QUERY, {
      variables: {
        handle: productHandle,
        selectedOptions,
        country: context.storefront.i18n.country,
        language: context.storefront.i18n.language,
      },
    });

    shop = resp.shop;
    product = resp.product;
  } catch (e: any) {
    console.log(e.stack);
    throw e;
  }
```

What if by default we also `console.error(error.stack)` or something before throwing? And allow people to disable that by a config option when creating the client?

Reply:
Or maybe we could add a `try catch` wrapper in the Oxygen request handler, and do a log there before bubbling the error?

Reply:
>I mean, to get something useful in the production logs, everyone has to manually try catch their queries and mutations, which is really painful:

Isn't the "better way" for them to have an ErrorBoundary which allows them to handle the error and also log it out? I guess it's not something we can enforce on our end, but it's supposed to be "the way" to handle errors in Remix, right?

Reply:
Yeah, I'm just feeling sad about all the existing merchants that have routes without ErrorBoundaries. Our skeleton template has proper error boundaries, but our demostore doesn't yet. Maybe I'll go add them to the demostore 🤔

Reply:
>>Since storefront.ts doesn't have access to the local domain or port, what about simply adding the query/variables information to error.cause here, and then picking it up at the CLI level?
>
>I'm cool with that.

Actually, thinking about this more, and specifically in the context of what Bret and I are talking about above - this solution would require devs to _not_ handle errors in their ErrorBoundary, right? If they do handle the error (and decide to not log it), that would mean that it wouldn't be logged to their server terminal, right? 

Additionally, it seems that Remix messes around with errors (@blittle is exploring this) which potentially also means that maybe we can't go forward with this idea?